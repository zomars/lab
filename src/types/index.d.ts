declare module 'typography-theme-wikipedia'
