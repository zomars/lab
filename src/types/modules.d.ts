declare module '*.woff2';
declare module '*.svg';
