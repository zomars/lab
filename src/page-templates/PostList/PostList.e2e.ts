import {
  ComponentWrapper,
  IComponentWrapperArgs,
} from '../../../e2e-tests/components/component-wrapper.e2e';
import { PostPreview } from '../../components/PostPreview/PostPreview.e2e';

enum PostListSelector {
  host = '[data-testid~=PostList]',
  post = '[data-testid~=PostList-PostPreview]',
}

export class PostList extends ComponentWrapper {
  constructor(args: IComponentWrapperArgs) {
    super({
      hostSelector: PostListSelector.host,
      ...args,
    });
  }

  public async getAllPosts(): Promise<PostPreview[]> {
    const $posts = await this.getElements(PostListSelector.post);

    return $posts.map(($post) => {
      return new PostPreview({
        scopeElement: this.$host,
        hostElement: $post,
        page: this.$page,
      });
    });
  }
}
