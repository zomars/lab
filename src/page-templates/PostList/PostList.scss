.PostList {
  margin-top: var(--aml-global-layout-padding-md);
  margin-bottom: var(--aml-global-layout-padding-md);

  &-PostPreview {
    &:last-child {
      border-bottom: none;
    }
  }

  &-Paginator {
    margin: var(--aml-global-layout-padding-md);
  }
}
