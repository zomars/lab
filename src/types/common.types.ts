import { Node } from 'gatsby';
import { ReactNode } from 'react';
import { ImageDataLike } from 'gatsby-plugin-image';

export interface INodeFields {
  [key: string]: unknown,
}

interface INode extends Node {
  fields: INodeFields;
}

interface IBlogPostFrontmatter {
  title: string,
  date: string,
  tags: string[],
  coverImage?: {
    childImageSharp: ImageDataLike,
  },
  published?: boolean,
  updated?: string,
  summary?: string & ReactNode,
  description: string,
}

interface IReadingTime {
  text: string,
  words: number,
}

interface IBlogPostFields extends INodeFields {
  slug: string,
  readingTime: IReadingTime,
}

export interface IBlogPost extends INode {
  frontmatter: IBlogPostFrontmatter;
  fields: IBlogPostFields;
  excerpt?: string;
  body?: string;
}

export interface IUniquePostTag {
  name: string;
  count: number;
}

export interface ISiteMetadata {
  title: string;
  description: string;
  author: string;
  siteUrl: string;
}

export interface IReactNodeProps {
  children: string | ReactNode;
}

export interface IPostTagsMappingQResponse {
  allPosts: {
    posts: IBlogPost[],
  },
  uniqueTags: {
    tags: IUniquePostTag[],
  },
}
