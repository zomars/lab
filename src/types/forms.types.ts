export interface ISendMessageFormPayload {
  message: string;
  captcha: string;
}
