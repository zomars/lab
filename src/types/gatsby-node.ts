import { Node } from 'gatsby';

export interface INodeFields {
  slug: string,
  [key: string]: unknown,
}

export interface INode extends Node {
  fields: INodeFields;
}
