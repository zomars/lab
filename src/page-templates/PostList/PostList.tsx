import React, { ReactElement } from 'react';
import { graphql, PageRendererProps } from 'gatsby';
import { cn } from '@bem-react/classname';

import { Layout } from '../../components/Layout';
import { Seo } from '../../components/Seo';
import { PostPreview } from '../../components/PostPreview/PostPreview';
import { IBlogPost } from '../../types/common.types';
import { postsPerPage } from '../../constants';
import { PostListPaginator } from './PostListPaginator/PostListPaginator';

import './PostList.scss';

export const postsQuery = graphql`
  query blogListQuery($skip: Int!, $limit: Int!, $tag: String!) {
    onePagePosts: allMdx(
      filter: {
        frontmatter: {
          published: { ne: false }
          tags: { in: [$tag] }
        }
      }
      sort: { fields: [frontmatter___date], order: DESC }
      limit: $limit
      skip: $skip
    ) {
      totalCount
      posts: nodes {
        excerpt(pruneLength: 400)
        fields {
          slug,
          readingTime {
            text,
            words,
          }
        }
        frontmatter {
          date(formatString: "MMM DD, YYYY")
          updated(formatString: "MMM DD, YYYY")
          tags
          title
          coverImage {
            childImageSharp {
              gatsbyImageData(height: 300)
            }
          }
        }
      }
    }
  }
`;

interface IGqlResponse {
  onePagePosts: {
    posts: IBlogPost[],
    totalCount: number,
  },
}

interface IPostListPageContext {
  tag: string,
  currentPage: number,
}

interface IPostListProps extends PageRendererProps {
  data: IGqlResponse;
  pageContext: IPostListPageContext;
}

const keywords = [
  'blog',
  'gatsby',
  'javascript',
  'react',
];

const cnPostList = cn('PostList');

export function PostList(props: IPostListProps): ReactElement {
  const { data, pageContext } = props;
  const { posts, totalCount } = data.onePagePosts;

  const postElements = posts.map(
    (post: IBlogPost) => (
      <PostPreview
        className = { cnPostList('PostPreview') }
        tag = { pageContext.tag }
        post = { post }
        key = { post.fields.slug }
      />
    )
  );

  const numPages = Math.ceil(totalCount / postsPerPage);

  return (
    <Layout>
      <Seo
        title = { `All #${ pageContext.tag } posts, page ${ pageContext.currentPage }` }
        keywords = { keywords }
      />

      { postElements }

      <PostListPaginator
        className = { cnPostList('Paginator') }
        current = { pageContext.currentPage }
        length = { numPages }
        tag = { pageContext.tag }
      />
    </Layout>
  );
}

// eslint-disable-next-line import/no-default-export
export default PostList;
