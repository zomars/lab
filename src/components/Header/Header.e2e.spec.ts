import { IGlobal } from '../../../e2e-test/e2e.types';
import { PostList } from '../../page-templates/PostList/PostList.e2e';
import {
  Header,
  MenuListItem,
  menuLabelsMap,
} from './Header.e2e';

const localGlobal = global as IGlobal & typeof globalThis;

describe('site header', () => {
  let header: Header = null;

  beforeEach(async () => {
    await jestPlaywright.resetContext();

    context.setDefaultTimeout(localGlobal.defaultTimeout);

    await page.goto(localGlobal.url);

    header = new Header();
  });

  it('is present on index page', async () => {
    await expect(header.isConnected).resolves.toBe(true);

    const promises = Array.from(
      menuLabelsMap.entries(),
    )
      .map(async ([key, label]: [MenuListItem, string]) => {
        const $element = await header.getMenuItem(key);

        return expect($element).toHaveText(label);
      });

    await Promise.all(promises);
  });

  it('click navigates user between pages', async () => {
    await expect(header.isConnected).resolves.toBe(true);

    await expect(header.clickMenuItem(MenuListItem.sources)).resolves.toBe(true);

    await expect(page).toHaveText(
      'h2',
      'Blogs',
    );

    await expect(header.clickMenuItem(MenuListItem.techPosts)).resolves.toBe(true);

    const techPostList = new PostList();

    await expect(techPostList.isConnected).resolves.toBe(true);

    const allTechPostsTags = await techPostList.getAllPostsTags();

    const techTags = allTechPostsTags.filter(tag => tag === 'tech');

    expect(techTags.length).toBeGreaterThanOrEqual(3);

    await expect(header.clickMenuItem(MenuListItem.carPosts)).resolves.toBe(true);

    await expect(techPostList.isConnected).resolves.toBe(false);

    const carPostList = new PostList();

    await expect(carPostList.isConnected).resolves.toBe(true);

    const allCarPostsTags = await carPostList.getAllPostsTags();

    const carTags = allCarPostsTags.filter(tag => tag === 'cars');

    // seems to be affected by the race condition (dev build)
    expect(carTags.length).toBeGreaterThanOrEqual(4);

    await expect(header.clickMenuItem(MenuListItem.about)).resolves.toBe(true);

    await expect(carPostList.isConnected).resolves.toBe(false);

    await expect(page).toHaveText(
      'h1',
      'About me',
    );
  });

  it('highlights selected tab only', async () => {
    await expect(header.isConnected).resolves.toBe(true);

    await expect(header.isSelected(MenuListItem.techPosts)).resolves.toBe(true);

    await expect(header.isSelected(MenuListItem.carPosts)).resolves.toBe(false);
  });
});
