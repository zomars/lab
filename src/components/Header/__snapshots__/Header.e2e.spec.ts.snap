// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`site header click navigates user between pages 1`] = `
Array [
  "How to Wait for Single Page Navigation and Re-hydration with Playwright and React Router",
  "Invalidate or Clear Github Actions Cache",
  "Github Actions for Dependabot PRs",
  "Optimize GitHub Actions for UI Builds and Tests with Cache",
  "The Making of a Manager by Julie Zhuo",
  "Efficient Code Review",
  "Good Parts of AngularJs Framework",
  "Unglamorous end of AngularJs development",
]
`;

exports[`site header click navigates user between pages 2`] = `
Array [
  "Audi B8 S4/S5/TT Seats into MK6 GTI Installation Guide",
]
`;
